File: e2e_tests/helpers/ab_update.py
#!/usr/bin/env python3
import argparse
import io
import time
from typing import Dict, List, Tuple
import yaml
import logging
from invoke.exceptions import CommandTimedOut
from ssh_utilities import (
    LOCAL_TRIDENT_CONFIG_PATH,
    create_ssh_connection,
    run_ssh_command,
    trident_run,
)

RETRY_INTERVAL = 60
MAX_RETRIES = 5

logging.basicConfig(level=logging.INFO)
log = logging.getLogger("ab_update")


class YamlSafeLoader(yaml.SafeLoader):
    def accept_image(self, node):
        return self.construct_mapping(node)


def trident_run_command(
    connection,
    runtime_env,
    stage_ab_update,
    finalize_ab_update,
    trident_config_path,
):
    """
    Composes and runs a command to trigger A/B update on the host and processes a range of
    possible outcomes.
    """

    # Determine the allowed operations to set
    allowed_operations = []
    if stage_ab_update:
        print("Staging of A/B update requested. Adding 'stage' to allowed operations.")
        allowed_operations.append("stage")
    if finalize_ab_update:
        print(
            "Finalizing of A/B update requested. Adding 'finalize' to allowed operations."
        )
        allowed_operations.append("finalize")
    allowed_operations_str = ",".join(allowed_operations)

    for attempt in range(MAX_RETRIES):
        try:
            print(f"Attempt {attempt + 1}: Running Trident run command...")

            # Provide -c arg, the full path to the RW Trident config.
            trident_return_code, trident_stdout, trident_stderr = trident_run(
                connection,
                f"run -v trace -c {trident_config_path} --allowed-operations {allowed_operations_str}",
                runtime_env,
            )

            if (
                trident_return_code == 0
                and "Staging of update 'AbUpdate' succeeded" in trident_stderr
            ):
                print("Staging of A/B update succeeded")
                return
            elif trident_return_code == -1 and "Rebooting system" in trident_stderr:
                print("Host rebooted successfully")
                return
            elif trident_return_code == 2 and (
                "Failed to run post-configure script 'fail-on-the-first-run'"
                in trident_stderr
            ):
                print("Detected intentional failure. Re-running...")
                continue
            else:
                raise Exception(
                    f"Unexpected exit code {trident_return_code}. Output: {trident_stdout + trident_stderr}"
                )
        except Exception as e:
            print(f"Exception during Trident run: {e}")
            raise
    raise Exception("Maximum retries exceeded for Trident run")


def update_osimage_url(runtime_env, destination_directory, host_config, version):
    """Updates the OS image URL in Host Configuration."""

    host_directory = "/" if runtime_env == "host" else "/host/"
    destination_directory = destination_directory.strip("/")

    old_url = host_config.get("osImage", {}).get("url", None)
    if old_url is None:
        raise Exception("No current osImage URL found in Host Configuration")

    if old_url.startswith("http://"):
        # Extract part between the last '/' and '.cosi'
        image_name = old_url.split("/")[-1].split(".cosi")[0]
    else:
        # Additional logic to strip "_vN", e.g. for URLs like file:///abupdate/regular_v2.cosi,
        # extract 'regular'; for file:///run/verity_v2.cosi, extract 'verity'
        image_name = "_".join(
            old_url.split("/")[-1].split(".cosi")[0].rsplit("_", 1)[:-1]
        )
    new_url = (
        f"file://{host_directory}{destination_directory}/{image_name}_v{version}.cosi"
    )

    log.info(f"Updating OS image URL from {old_url} to {new_url}")
    host_config["osImage"]["url"] = new_url

    # Remove selfUpgrade field from the Trident config
    host_config["trident"]["selfUpgrade"] = False


def trigger_ab_update(
    ip_address,
    user_name,
    keys_file_path,
    destination_directory,
    trident_config_path,
    version,
    runtime_env,
    stage_ab_update,
    finalize_ab_update,
):
    """Connects to the host via SSH, updates the local Trident config, and re-runs Trident"""
    # Set up SSH client
    connection = create_ssh_connection(ip_address, user_name, keys_file_path)

    _, trident_stdout, _ = trident_run(connection, "get", runtime_env)
    host_status = trident_stdout.strip()

    YamlSafeLoader.add_constructor("!image", YamlSafeLoader.accept_image)
    host_status_dict = yaml.load(host_status, Loader=yaml.SafeLoader)
    host_config = host_status_dict["spec"]

    # If staging of A/B update is required, update the OS image URL in Host Configuration
    if stage_ab_update:
        update_osimage_url(runtime_env, destination_directory, host_config, version)
        connection.run("sudo mkdir -p /tmp/staging")
        connection.run("sudo chmod 777 /tmp/staging")
        connection.put(io.StringIO(yaml.dump(host_config)), "/tmp/staging/hc.yaml")
        connection.run(f"sudo mv /tmp/staging/hc.yaml {trident_config_path}")

    # Re-run Trident and capture logs
    print("Re-running Trident to trigger A/B update", flush=True)
    trident_run_command(
        connection,
        runtime_env,
        stage_ab_update,
        finalize_ab_update,
        trident_config_path,
    )
    connection.close()

    # For container testing, finalize the A/B update by manually triggering
    # a Trident run on the updated Runtime OS.
    if finalize_ab_update and runtime_env == "container":
        for attempt in range(MAX_RETRIES):
            try:
                time.sleep(RETRY_INTERVAL)
                # Re-establish connection
                print(f"Attempt {attempt + 1}: Run Trident after A/B update")

                connection = create_ssh_connection(
                    ip_address, user_name, keys_file_path
                )

                trident_return_code, trident_stdout, trident_stderr = trident_run(
                    connection, f"run", runtime_env
                )

                if trident_return_code == 0:
                    break
                else:
                    raise Exception(trident_stdout + trident_stderr)
            except Exception as e:
                print(f"Trident run attempt {attempt + 1} failed: {e}")
            finally:
                connection.close()
        else:
            raise Exception(
                "Maximum attempts exceeded. Trident was unable to successfully run after A/B update."
            )

    return


def main():
    # Setting argument_default=argparse.SUPPRESS means that the program will
    # halt attribute creation if no values provided for arg-s
    parser = argparse.ArgumentParser(
        allow_abbrev=True, argument_default=argparse.SUPPRESS
    )
    parser.add_argument(
        "-i",
        "--ip-address",
        type=str,
        help="IP address of the host.",
    )
    parser.add_argument(
        "-u",
        "--user-name",
        type=str,
        help="User name to ssh into the host with.",
    )
    parser.add_argument(
        "-k",
        "--keys-file-path",
        type=str,
        help="Full path to the file containing the host ssh keys.",
    )
    parser.add_argument(
        "-d",
        "--destination-directory",
        type=str,
        help="Read-write directory on the host that contains the runtime OS images for the A/B update.",
    )
    parser.add_argument(
        "-c",
        "--trident-config",
        type=str,
        help="File name of the custom read-write Trident config on the host to point Trident to.",
    )
    parser.add_argument(
        "-v",
        "--version",
        type=str,
        help="Version of runtime OS images for the A/B update.",
    )
    parser.add_argument(
        "-e",
        "--runtime-env",
        action="store",
        type=str,
        choices=["host", "container"],
        default="host",
        help="Runtime environment for trident: 'host' or 'container'. Default is 'host'.",
    )
    parser.add_argument(
        "-s",
        "--stage-ab-update",
        action="store_true",
        help="Controls whether A/B update should be staged.",
    )
    parser.add_argument(
        "-f",
        "--finalize-ab-update",
        action="store_true",
        help="Controls whether A/B update should be finalized.",
    )

    args = parser.parse_args()

    # Set to False if flags not provided
    stage_ab_update = getattr(args, "stage_ab_update", False)
    finalize_ab_update = getattr(args, "finalize_ab_update", False)

    # Call helper func that triggers A/B update
    trigger_ab_update(
        args.ip_address,
        args.user_name,
        args.keys_file_path,
        args.destination_directory,
        args.trident_config,
        args.version,
        args.runtime_env,
        stage_ab_update,
        finalize_ab_update,
    )


if __name__ == "__main__":
    main()

File: storm/suites/trident/helpers/ab_update.go
package helpers

import (
	"fmt"
	"path"
	"regexp"
	"storm/pkg/storm"
	"storm/suites/trident/utils"
	"strings"
	"time"

	"golang.org/x/crypto/ssh"
	"gopkg.in/yaml.v3"
)

type AbUpdateHelper struct {
	args struct {
		utils.SshCliSettings `embed:""`
		utils.EnvCliSettings `embed:""`
		DestinationDirectory string `short:"d" required:"" help:"Read-write directory on the host that contains the runtime OS images for the A/B update."`
		TridentConfig        string `short:"c" required:"" help:"File name of the custom read-write Trident config on the host to point Trident to."`
		Version              string `short:"v" required:"" help:"Version of the Trident image to use for the A/B update."`
		StageAbUpdate        bool   `short:"s" help:"Controls whether A/B update should be staged."`
		FinalizeAbUpdate     bool   `short:"f" help:"Controls whether A/B update should be finalized."`
	}

	client *ssh.Client
	config map[string]interface{}
}

func (h AbUpdateHelper) Name() string {
	return "ab-update"
}

func (h *AbUpdateHelper) Args() any {
	return &h.args
}

func (h *AbUpdateHelper) RegisterTestCases(r storm.TestRegistrar) error {
	r.RegisterTestCase("get-config", h.getHostConfig)
	r.RegisterTestCase("update-hc", h.updateHostConfig)
	r.RegisterTestCase("trigger-update", h.triggerTridentUpdate)
	r.RegisterTestCase("check-trident-service", h.checkTridentService)
	return nil
}

func (h *AbUpdateHelper) getHostConfig(tc storm.TestCase) error {
	if h.args.Env == utils.TridentEnvironmentNone {
		return fmt.Errorf("environment %s is not supported", h.args.Env)
	}

	var err error
	h.client, err = utils.OpenSshClient(h.args.SshCliSettings)
	if err != nil {
		tc.Error(err)
	}

	tc.SuiteCleanup(func() {
		if h.client != nil {
			h.client.Close()
		}
	})

	out, err := utils.InvokeTrident(h.args.Env, h.client, "get configuration")
	if err != nil {
		return fmt.Errorf("failed to invoke Trident: %w", err)
	}

	if err := out.Check(); err != nil {
		return fmt.Errorf("failed to run trident to get host config: %w", err)
	}

	tc.Logger().Debugf("Trident stdout:\n%s", out.Stdout)
	tc.Logger().Debugf("Trident stderr:\n%s", out.Stderr)

	err = yaml.Unmarshal([]byte(out.Stdout), &h.config)
	if err != nil {
		return fmt.Errorf("failed to unmarshal YAML: %w", err)
	}
	tc.Logger().Infof("Trident configuration: %v", h.config)

	return nil
}

func (h *AbUpdateHelper) updateHostConfig(tc storm.TestCase) error {
	if !h.args.StageAbUpdate {
		tc.Skip("Staging not requested")
	}

	// Extract the OLD URL from the configuration
	oldUrl, ok := h.config["image"].(map[string]interface{})["url"].(string)
	if !ok {
		return fmt.Errorf("failed to get old image URL from configuration")
	}

	tc.Logger().Infof("Old image URL: %s", oldUrl)

	base := path.Base(oldUrl)

	matches := regexp.MustCompile(`^(.*?)(_v\d+)?\.(.+)$`).FindStringSubmatch(base)

	if len(matches) != 4 {
		return fmt.Errorf("failed to parse image name: %s", base)
	}

	name := matches[1]
	ext := matches[3]

	newCosiName := fmt.Sprintf("%s_v%s.%s", name, h.args.Version, ext)
	newCosiPath := path.Join(h.args.DestinationDirectory, newCosiName)
	tridentCosiPath := path.Join(h.args.Env.HostPath(), newCosiPath)
	newUrl := fmt.Sprintf("file://%s", tridentCosiPath)
	tc.Logger().Infof("New image URL: %s", newUrl)

	// Update the image URL in the configuration
	h.config["image"].(map[string]any)["url"] = newUrl

	// Set the config to NOT self-upgrade
	trident, ok := h.config["trident"].(map[string]any)
	if !ok {
		trident = make(map[string]any)
		h.config["trident"] = trident
	}

	trident["selfUpgrade"] = false

	// Delete the storage section from the config, not needed for A/B update
	delete(h.config, "storage")

	hc_yaml, err := yaml.Marshal(h.config)
	if err != nil {
		return fmt.Errorf("failed to marshal YAML: %w", err)
	}

	sftpClient, err := utils.NewSftpSudoClient(h.client)
	if err != nil {
		return fmt.Errorf("failed to create SudoSFTP client: %w", err)
	}
	defer sftpClient.Close()

	// Ensure the cosi file exists
	tc.Logger().Infof("Checking if new COSI file exists at %s", newCosiPath)
	_, err = sftpClient.Stat(newCosiPath)
	if err != nil {
		fmt.Println("Yielding to the error")
		return fmt.Errorf("failed to stat new COSI file at %s: %w", newCosiPath, err)
	}

	err = sftpClient.MkdirAll(path.Dir(h.args.TridentConfig))
	if err != nil {
		return fmt.Errorf("failed to create directory for new Host Config file: %w", err)
	}

	file, err := sftpClient.Create(h.args.TridentConfig)
	if err != nil {
		return fmt.Errorf("failed to create new Host Config file: %w", err)
	}
	defer file.Close()

	_, err = file.Write(hc_yaml)
	if err != nil {
		return fmt.Errorf("failed to write new Host Config file: %w", err)
	}

	err = file.Chmod(0644)
	if err != nil {
		return fmt.Errorf("failed to change permissions of new Host Config file: %w", err)
	}

	err = file.Chown(0, 0)
	if err != nil {
		return fmt.Errorf("failed to change ownership of new Host Config file: %w", err)
	}

	return nil
}

func (h *AbUpdateHelper) triggerTridentUpdate(tc storm.TestCase) error {
	allowedOperations := make([]string, 0)

	if h.args.StageAbUpdate {
		tc.Logger().Infof("Allowed operations: stage")
		allowedOperations = append(allowedOperations, "stage")
	}

	if h.args.FinalizeAbUpdate {
		tc.Logger().Infof("Allowed operations: finalize")
		allowedOperations = append(allowedOperations, "finalize")
	}

	args := fmt.Sprintf(
		"update -v trace %s --allowed-operations %s",
		path.Join(h.args.Env.HostPath(), h.args.TridentConfig),
		strings.Join(allowedOperations, ","),
	)

	file, err := utils.CommandOutput(h.client, tc.Logger(), fmt.Sprintf("sudo cat %s", h.args.TridentConfig))
	if err != nil {
		return fmt.Errorf("failed to read new Host Config file: %w", err)
	}

	tc.Logger().Debugf("Trident config file:\n%s", file)

	for i := 1; ; i++ {
		tc.Logger().Infof("Invoking Trident attempt #%d with args: %s", i, args)

		out, err := utils.InvokeTrident(h.args.Env, h.client, args)
		if err != nil {
			if err, ok := err.(*ssh.ExitMissingError); ok && strings.Contains(out.Stderr, "Rebooting system") {
				// The connection closed without an exit code, and the output contains "Rebooting system".
				// This indicates that the host has rebooted.
				tc.Logger().Infof("Host rebooted successfully")
				break
			} else {
				// Some unknown error occurred.
				tc.Logger().Errorf("Failed to invoke Trident: %s; %s", err, out.Report())
				return fmt.Errorf("failed to invoke Trident: %w", err)
			}
		}

		if out.Status == 0 && strings.Contains(out.Stderr, "Staging of update 'AbUpdate' succeeded") {
			tc.Logger().Infof("Staging of update 'AbUpdate' succeeded")
			break
		}

		if out.Status == 2 && strings.Contains(out.Stderr, "Failed to run post-configure script 'fail-on-the-first-run'") {
			tc.Logger().Infof("Detected intentional failure. Re-running...")
			continue
		}

		tc.Logger().Errorf("Trident update failed %s", out.Report())

		tc.Fail(fmt.Sprintf("Trident update failed with status %d", out.Status))
	}

	// On success close the client because the host will reboot into the new OS.
	h.client.Close()
	h.client = nil

	return nil
}

func (h *AbUpdateHelper) checkTridentService(tc storm.TestCase) error {
	if h.args.Env == utils.TridentEnvironmentNone {
		tc.Skip("No Trident environment specified")
	}

	tc.Logger().Infof("Waiting for the host to reboot and come back online...")
	time.Sleep(time.Second * 10)

	// Reconnect via SSH to the updated OS
	_, err := utils.Retry(
		h.args.TimeoutDuration(),
		time.Second*5,
		func(attempt int) (*bool, error) {
			tc.Logger().Infof("SSH dial to '%s' (attempt %d)", h.args.SshCliSettings.FullHost(), attempt)
			client, err := utils.OpenSshClient(h.args.SshCliSettings)
			if err != nil {
				tc.Logger().Warnf("Failed to dial SSH server '%s': %s", h.args.SshCliSettings.FullHost(), err)
				return nil, err
			}
			defer client.Close()

			tc.Logger().Infof("SSH dial to '%s' succeeded", h.args.SshCliSettings.FullHost())

			err = utils.CheckTridentService(client, tc.Logger(), h.args.Env, h.args.TimeoutDuration())
			if err != nil {
				tc.Logger().Warnf("Trident service is not in expected state: %s", err)
				return nil, err
			}

			tc.Logger().Infof("Trident service is in expected state")
			return nil, nil
		},
	)
	if err != nil {
		// Log this as a test failure
		tc.FailFromError(err)
	}

	return nil
}
